src/mysql_vm_helpers.py

```diff
@@ -131,32 +130,41 @@ def install_and_configure_mysql_dependencies() -> None:
         """Install and configure MySQL dependencies.
 
         Raises
-            subprocess.CalledProcessError: if issue updating apt or creating mysqlsh common dir
-            apt.PackageNotFoundError, apt.PackageError: if issue install mysql server
-            snap.SnapNotFOundError, snap.SnapError: if issue installing mysql shell snap
+            subprocess.CalledProcessError: if issue creating mysqlsh common dir
+            snap.SnapNotFoundError, snap.SnapError: if issue installing charmed-mysql snap
         """
+        logger.debug("Retrieving snap cache")
+        cache = snap.SnapCache()
+        charmed_mysql = cache[CHARMED_MYSQL_SNAP_NAME]
+        # This charm can override/use an existing snap installation only if the snap was previously
+        # installed by this charm.
+        # Otherwise, the snap could be in use by another charm (e.g. MySQL Router charm).
+        installed_by_mysql_server_file = pathlib.Path(
+            CHARMED_MYSQL_COMMON_DIRECTORY, "installed_by_mysql_server_charm"
+        )
+        if charmed_mysql.present and not installed_by_mysql_server_file.exists():
+            logger.error(
+                f"{CHARMED_MYSQL_SNAP_NAME} snap already installed on machine. Installation aborted"
+            )
+            raise Exception(
+                f"Multiple {CHARMED_MYSQL_SNAP_NAME} snap installs not supported on one machine"
+            )
+
         try:
             # install the charmed-mysql snap
-            logger.debug("Retrieving snap cache")
-            cache = snap.SnapCache()
-            charmed_mysql = cache[CHARMED_MYSQL_SNAP_NAME]
-
-            if not charmed_mysql.present:
-                logger.debug("Installing charmed-mysql snap")
-                charmed_mysql.ensure(snap.SnapState.Present, revision=CHARMED_MYSQL_SNAP_REVISION)
+            logger.debug("Installing charmed-mysql snap")
+            charmed_mysql.ensure(snap.SnapState.Present, revision=CHARMED_MYSQL_SNAP_REVISION)
```

Comment:
We have strict match charm revision == snap revision.
If we installing charm revision A, we must ensure snap revision A is installed, otherwise we will spend days to troubleshoot mismatch.